### PR TITLE
feat: SQL Editor Charts

### PR DESCRIPTION
**Pull request by jordienr**
## What kind of change does this PR introduce?

adds charts to SQL editor

## What is the current behavior?

no charts in SQL editor

## What is the new behavior?

charts in SQL editor

## Additional context

CleanShot 2024-02-28 at 17 13 25@2x

### CODE REVIEW

**Comment by saltcod** [comment]
screenshot-2024-02-28-at-14 18 14

**Comment by joshenlim** [comment]
could add a draggable marker here between the chart and the control panel (like how we do for the code editor and the utility panel)
image

**Comment by joshenlim** [comment]
Possible to show axis legends based on the selected columns? The x axis probably should be dynamic based on the column type (if possible) - e.g shouldnt be a date if the values are numbers

image

**Comment by joshenlim** [comment]
(on the fence) Am thinking the export options could be dynamic based on the tabs
results -> CSV, markdown, JSON
charts -> export as png? not sure what other forms might be useful

image

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityPanel.tsx [hunk 1]
```diff
@@ -19,22 +25,61 @@ const UtilityPanel = ({
...
-          hasSelection={hasSelection}
-          prettifyQuery={prettifyQuery}
-          executeQuery={executeQuery}
-        />
-      </div>
-
-      <div className="flex-1 p-0 pt-0 pb-0">
-        <UtilityTabResults id={id} isExecuting={isExecuting} />
+      <div className="h-full">
+        <Tabs_Shadcn_ defaultValue="results" className="w-full h-full">
```
Comment (MildTomato) [inline apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityPanel.tsx:47]: <div className="h-full ..
  <component className="h-full ..
  
  
suss...

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx [hunk 2]
```diff
@@ -105,18 +104,23 @@ const UtilityActions = ({
...
-              <div className="flex items-center space-x-1">
-                {os === 'macos' ? (
-                  <IconCommand size={10} strokeWidth={1.5} />
-                ) : (
-                  <p className="text-xs text-foreground-light">CTRL</p>
-                )}
-                <IconCornerDownLeft size={10} strokeWidth={1.5} />
-              </div>
+              isExecuting ? (
+                <div className="w-6 flex items-center justify-center">
```
Comment (MildTomato) [inline apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx:111]: this div doesn't do anything

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/ResultsDropdown.tsx [hunk 3]
```diff
@@ -136,6 +147,10 @@ const ResultsDropdown = ({ id, isExecuting }: ResultsDropdownProps) => {
             <IconClipboard size="tiny" />
             <p>Copy as markdown</p>
           </DropdownMenuItem>
+          <DropdownMenuItem onClick={onCopyAsJSON} className="space-x-2">
+            <IconClipboard size="tiny" />
+            <p>Copy as JSON</p>
```
Comment (MildTomato) [inline apps/studio/components/interfaces/SQLEditor/UtilityPanel/ResultsDropdown.tsx:152]: remove `<p>` - not required

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/ResultsDropdown.tsx [hunk 4]
```diff
@@ -136,6 +147,10 @@ const ResultsDropdown = ({ id, isExecuting }: ResultsDropdownProps) => {
             <IconClipboard size="tiny" />
             <p>Copy as markdown</p>
           </DropdownMenuItem>
+          <DropdownMenuItem onClick={onCopyAsJSON} className="space-x-2">
```
Comment (MildTomato) [inline apps/studio/components/interfaces/SQLEditor/UtilityPanel/ResultsDropdown.tsx:150]: `flex gap-2`

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityPanel.tsx (hunk 1 above)
Comment (MildTomato) [reply apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityPanel.tsx:47]: non blocker

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx (hunk 2 above)
Comment (MildTomato) [reply apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx:111]: non blocker

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/ResultsDropdown.tsx (hunk 3 above)
Comment (MildTomato) [reply apps/studio/components/interfaces/SQLEditor/UtilityPanel/ResultsDropdown.tsx:152]: non blocker

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/ResultsDropdown.tsx (hunk 4 above)
Comment (MildTomato) [reply apps/studio/components/interfaces/SQLEditor/UtilityPanel/ResultsDropdown.tsx:150]: non blocker

File: apps/studio/components/interfaces/SQLEditor/SQLEditor.tsx [hunk 5]
```diff
@@ -980,7 +985,11 @@ const SQLEditor = () => {
           <ResizableHandle withHandle />
           <ResizablePanel collapsible collapsedSize={10} minSize={20}>
             {isLoading ? (
-              <div className="flex h-full w-full items-center justify-center">Loading...</div>
+              <div className="flex h-full w-full items-center justify-center">
+                <Loading active={true}>
+                  <></>
+                </Loading>
+              </div>
```
Comment (MildTomato) [inline apps/studio/components/interfaces/SQLEditor/SQLEditor.tsx:992]: should we fix this component? this looks like terrible DX

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/ChartConfig.tsx [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+              <h2 className="text-sm text-foreground-lighter">Chart options</h2>
+              {config.xKey && config.yKey && (
+                <Button
+                  type="text"
+                  onClick={() => {
+                    const currentX = config.xKey
+                    const currentY = config.yKey
+                    onConfigChange({ ...config, xKey: currentY, yKey: currentX })
+                  }}
+                  title="Swap X and Y axis"
```
Comment (MildTomato) [inline apps/studio/components/interfaces/SQLEditor/UtilityPanel/ChartConfig.tsx:117]: nice

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/ChartConfig.tsx [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+                  {resultKeys.map((key) => (
+                    <SelectItem_Shadcn_ value={key} key={key}>
+                      {key}
+                    </SelectItem_Shadcn_>
+                  ))}
+                </SelectGroup_Shadcn_>
+              </SelectContent_Shadcn_>
+            </Select_Shadcn_>
+
+            <Checkbox
```
Comment (MildTomato) [inline apps/studio/components/interfaces/SQLEditor/UtilityPanel/ChartConfig.tsx:164]: use Checkbox_shadcn ?

This is a change I think is worth doing to reduce refactor later.

**Review by MildTomato — CHANGES_REQUESTED** [review]

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/ChartConfig.tsx [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    return 'string'
+  }
+
+  const xKeyDateFormat = getDateFormat(config.xKey)
+
+  return (
+    <ResizablePanelGroup direction="horizontal" className="flex-grow h-full">
+      <>
+        <ResizablePanel className="p-4" defaultSize={85}>
+          <div className="h-full">
```
Comment (MildTomato) [inline apps/studio/components/interfaces/SQLEditor/UtilityPanel/ChartConfig.tsx:90]: useless div?
move h-full to `<ResizablePanel/>`

**Review by ivasilov — APPROVED** [review]

**Review by MildTomato — APPROVED** [review]

**Comment by blechatellier** [comment]
Great stuff, any plan to be able to create a custom dashboard using multiple charts (like geckoboard)?

**Comment by zedeus** [comment]
Any plans to add this to the log explorer?

**Pull request merged**